File: ios/SIOPKit/Sources/SIOPKit/AuthorizationRequest.swift
import Foundation

/// Self-Issued OpenID Provider Request (OpenID Connect Core 1.0 Section 7.3).
public struct AuthorizationRequest: Equatable {
    public let responseType: String
    public let scope: [String]
    /// Section 7.2: the client_id is the RP's redirect URI.
    public let clientID: String
    public let nonce: String
    public let state: String?
    public let idTokenHint: String?
    /// Raw JSON of the claims parameter, if any (Section 5.5).
    public let claims: String?
    /// Raw JSON of the registration parameter, if any (Section 7.2.1).
    public let registration: String?

    public init(url: URL) throws {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            throw SIOPError.invalidRequest("malformed URL")
        }
        var params: [String: String] = [:]
        for item in components.queryItems ?? [] {
            params[item.name] = item.value
        }

        guard let responseType = params["response_type"] else {
            throw SIOPError.invalidRequest("response_type is required")
        }
        guard responseType == "id_token" else {
            throw SIOPError.unsupportedResponseType(responseType)
        }

        let scope = (params["scope"] ?? "").split(separator: " ").map(String.init)
        guard scope.contains("openid") else {
            throw SIOPError.invalidScope
        }

        guard let clientID = params["client_id"], !clientID.isEmpty else {
            throw SIOPError.invalidRequest("client_id is required")
        }
        guard let redirectURL = URL(string: clientID), redirectURL.scheme != nil else {
            throw SIOPError.invalidRequest("client_id must be the RP redirect URI")
        }
        if let redirectURI = params["redirect_uri"], redirectURI != clientID {
            throw SIOPError.invalidRequest("redirect_uri must equal client_id")
        }

        // response_type=id_token is the implicit flow, so nonce is required (Section 3.2.2.1).
        guard let nonce = params["nonce"], !nonce.isEmpty else {
            throw SIOPError.invalidRequest("nonce is required")
        }

        self.responseType = responseType
        self.scope = scope
        self.clientID = clientID
        self.nonce = nonce
        self.state = params["state"]
        self.idTokenHint = params["id_token_hint"]
        self.claims = params["claims"]
        self.registration = params["registration"]
    }
}

File: android/siopkit/src/main/kotlin/jp/co/pendako/siop/AuthorizationRequest.kt
package jp.co.pendako.siop

import java.net.URI

/** Self-Issued OpenID Provider Request (OpenID Connect Core 1.0 Section 7.3). */
data class AuthorizationRequest(
    val responseType: String,
    val scope: List<String>,
    /** Section 7.2: the client_id is the RP's redirect URI. */
    val clientId: String,
    val nonce: String,
    val state: String?,
    val idTokenHint: String?,
    /** Raw JSON of the claims parameter, if any (Section 5.5). */
    val claims: String?,
    /** Raw JSON of the registration parameter, if any (Section 7.2.1). */
    val registration: String?,
) {
    companion object {
        fun parse(url: String): AuthorizationRequest {
            val params = queryParameters(url)

            val responseType = params["response_type"]
                ?: throw SiopError.InvalidRequest("response_type is required")
            if (responseType != "id_token") throw SiopError.UnsupportedResponseType(responseType)

            val scope = (params["scope"] ?: "").split(" ").filter { it.isNotEmpty() }
            if (!scope.contains("openid")) throw SiopError.InvalidScope()

            val clientId = params["client_id"]?.takeIf { it.isNotEmpty() }
                ?: throw SiopError.InvalidRequest("client_id is required")
            runCatching { URI(clientId) }.getOrNull()?.scheme
                ?: throw SiopError.InvalidRequest("client_id must be the RP redirect URI")
            params["redirect_uri"]?.let {
                if (it != clientId) throw SiopError.InvalidRequest("redirect_uri must equal client_id")
            }

            // response_type=id_token is the implicit flow, so nonce is required
            // (Section 3.2.2.1).
            val nonce = params["nonce"]?.takeIf { it.isNotEmpty() }
                ?: throw SiopError.InvalidRequest("nonce is required")

            return AuthorizationRequest(
                responseType = responseType,
                scope = scope,
                clientId = clientId,
                nonce = nonce,
                state = params["state"],
                idTokenHint = params["id_token_hint"],
                claims = params["claims"],
                registration = params["registration"],
            )
        }

        private fun queryParameters(url: String): Map<String, String> {
            val query = url.substringAfter('?', "").substringBefore('#')
            if (query.isEmpty()) throw SiopError.InvalidRequest("no query parameters")
            return query.split("&").mapNotNull { pair ->
                if (pair.isEmpty()) return@mapNotNull null
                val name = pair.substringBefore('=')
                val value = pair.substringAfter('=', "")
                name.percentDecoded() to value.percentDecoded()
            }.toMap()
        }

        private fun String.percentDecoded(): String =
            java.net.URLDecoder.decode(this, Charsets.UTF_8)
    }
}
